stats/cloud/collector.go

```diff
@@ -54,9 +55,10 @@ type Collector struct {
 	anonymous bool
 	runStatus lib.RunStatus
 
-	bufferMutex      sync.Mutex
```

Comment:
As with in the other outputs we prefer if this is more explicitly named. For me if I have just a mutex called `mu` in a struct I would expect it's protecting the whole struct which .. clearly not the case. 

IMHO this will look a lot better if we refactor it (#1075)  so the whole logic about how we buffer samples is in one place and is just reused in all of the outputs as mentioned in https://github.com/loadimpact/k6/pull/1114#discussion_r313831931 . But until then I prefer if it's more explicit

Reply:
Correct me if I’m wrong, if you do call c.buffetMutex.Lock(), it does lock the whole struct, no?

Reply:
Well because k6 is in golang and [not rust](https://doc.rust-lang.org/std/sync/struct.Mutex.html)  mutexes do nothing, but lock themselves when you lock them :) There is absolutely nothing stopping anyone (in most languages) from not locking the mutex somewhere and using the variable that it is suppose to "protect". It will just lead to your program being wrong, but it will still compile and run :) 
So if you have a mutex and seems like it should be used for everything in a struct but it isn't it will look like there is a bug ;)

Reply:
I don’t agree. If you named it bufferLock, then when you lock it, one will expect that the lock won’t affect other fields. That will cause a bug.

You can find many places in Go standard library, which use the same pattern as I did in this PR. Give struct a mutex, document the fields it will protect.

Also, if we keep the bufferLock, then if I want protect stopSendingMetrics, I have to add another lock to make code meaningful, because I should’n use bufferLock to protect stopSendingMetrics

Reply:
> I don’t agree. If you named it bufferLock, then when you lock it, one will expect that the lock won’t affect other fields. That will cause a bug.

But... it won't affect other fields, so, it won't actually cause a bug. As @MStoykov said, Go mutexes don't actually lock any fields, they only [lock themselves](https://golang.org/pkg/sync/#Mutex.Lock). And it's up to the developer to use that property of theirs (i.e. **mut**ual **ex**clusion) to handle the synchronization of their code... So, as you can see, currently the `bufferMutex` is only used for the synchronization of access to the `bufferSamples` and `bufferHTTPTrails`, so `bufferMutex` seems a much more appropriate and self-documenting name than the super-generic `mu`... Theoretically if we want to (and if it won't cause any deadlocks, which I'm not sure), we could probably even slightly optimize the code by having 2 different mutexes here - one `bufferSampleMutex` for `bufferSamples` and another `bufferHTTPTrailMutex` for `bufferHTTPTrails`... 

> You can find many places in Go standard library, which use the same pattern as I did in this PR. Give struct a mutex, document the fields it will protect.

I'd be interested to see any examples you think are similar. In my head, it's better to have a self-documenting and descriptive property name, than to have a ultra-generic property name that you then need to document with some comment. Especially when you consider that this is not a simple data struct, rather it's a 500+ LoC file.

> Also, if we keep the bufferLock, then if I want protect stopSendingMetrics, I have to add another lock to make code meaningful, because I should’n use bufferLock to protect stopSendingMetrics

Yes, there's no reason I can see that we should reuse the same mutex for the buffers and for `stopSendingMetrics`. Also, the current implementation with a `bool` `stopSendingMetrics` doesn't seem ideal. You could use a simple atomic and then you wouldn't need any locking. 

Or better yet, you probably should handle this with a channel of some sort, that you can `close()` when you want to stop accepting new metrics. That way you can add a check for it it in the `select{}` clauses in `Run()` and you can return from that function, and you can also check for it in `Collect()` and not accept any new metrics.